### PR TITLE
Native ReadableStream API for Chat (+ more)

### PR DESCRIPTION
**Pull request by Gregoor**
Fixes https://linear.app/commonbase/issue/CB-93/chat-sdk-missing-features

...but atm session resuming is not quite working, namely the spec I got says
> client sends info message, if initialized -> chatting can begin
if not, send init message, wait for init response -> start chatting

And I realized that I don't quite know what "if not" means. Can you give me a hint, or an impl help @christophwitzko?

### CODE REVIEW

**Comment by christophwitzko** [comment]
Sorry for being so unclear, here is a more elaborated explanation:
The first message that the client should send is a `info` message. If the session was already initialized we get the following response:
``` javascript
{
  type: 'info',
  sessionId: 'xxx', // the session id which we used to established the websocket connection (/chats/<sessionId>)
  initialized: 1680955834113, // the timestamp when the session was initialized
  contextSize: 1 // the amount of messages in the context
}
```

The alternative response is when the session is new or was reset:
``` javascript
{
  type: 'info',
  error: 'session not initialized',
  sessionId: 'xxx', // the session id should be saved as a private member to re-establish the connection (I think you are already doing this)
  initialized: 0,
}
```
If so, we have to send the `init` message and start the session.

**Comment by Gregoor** [comment]
Thanks for the clarification! One more q: does it also mean users can only set vars+sessionData on session creation?
Then I guess I should change the `ChatClients` API to
```ts
type ChatClientOptions = { sessionId } | { variables }
```

**Comment by Gregoor** [comment]
Session resuming is now in! Thanks again for the hand.

Unfortunately things spiraled a bit, complexity-wise, for our integrators (which makes me question some things):
- the example now includes a kind-of-incorrect async `useEffect` (`useChatClient`) which uses an out-of-date update function to start the chat client
  - otherwise the first chat-connect attempt gets killed (by react double-running effects in development) and the second one fails because of the lock of the first one not being lifted yet
- previously it worked because I only started the chat session when the first message was sent, thus not running into the lock
  - now we do need to start the session immediately, so that we get the sessionId, for persistance

Now I'm wondering: isn't the lock in the wrong place? We don't want two sessions to send a message simultaneously, but they could still be connected at the same time no?

More generally, I'm once again wondering if we want to manage, and expose, the extra-websocket burden if what we're only doing it for locking behavior? As alluded to before the SDK code is mostly trying to hide the socket-state-machine behind a request-response model.

**Comment by Gregoor** [comment]
I am going to take this back into the workshop based on the comments I got in the call

**Comment by Gregoor** [comment]
Oh yeah @christophwitzko, the reason I could not e.g. stash the ChatClient initialization into a useState-initializer or a useMemo is because I made it async. I originally did that so that I have sync-access to the sessionId (either the give, or the newly created one)
I guess I made the wrong choice there and should rather just make that access async, then I can use state or memo.

**Comment by Gregoor** [comment]
Alrighty, I'm happy enough again, and opening this up to reviews

short video showing session resuming:

https://user-images.githubusercontent.com/4051932/231190875-ae557cb6-9629-4a19-ad95-6da997e9caf6.mov

File: chat-client.ts [hunk 1]
```diff
@@ -1,144 +1,163 @@
...
+          if (msg.error && msg.initialized !== 0) {
             reject(msg.error);
           } else {
-            resolve(wsClient);
+            const sessionId =
+              msg.sessionId ?? ("sessionId" in options && options.sessionId);
+            if (sessionId) {
+              resolve([ws, sessionId]);
+            } else {
+              reject("Missing sessionId");
```
Comment (christophwitzko) [inline chat-client.ts]: wouldn't it be better to use an Error object here:
`reject(new Error("Missing sessionId"));`

File: chat-client.ts [hunk 2]
```diff
@@ -1,144 +1,163 @@
...
 
-import { ClientOptions } from "./types";
-
-const initializeWSClient = memoizeOne(async (projectId: string) => {
-  return new Promise<WebSocket>((resolve, reject) => {
-    const wsClient = new WebSocket("wss://api.commonbase.com/chats");
+async function startWebSocketSession(options: ChatClientOptions) {
+  return new Promise<[WebSocket, string]>((resolve, reject) => {
+    const ws = new WebSocket(
+      "wss://api.commonbase.com/chats" +
```
Comment (heikir) [inline chat-client.ts]: Should this be a variable?

File: chat-client.ts [hunk 3]
```diff
@@ -1,144 +1,163 @@
...
 
     function cleanup() {
-      wsClient.removeEventListener("error", handleError);
-      wsClient.removeEventListener("message", handleMessage);
+      ws.removeEventListener("error", handleError);
+      ws.removeEventListener("message", handleMessage);
     }
 
-    const handleError = (error: Event) => {
+    const handleError = (error: unknown) => {
```
Comment (heikir) [inline chat-client.ts]: This cannot be an error event type?

File: chat-client.ts [hunk 4]
```diff
@@ -1,144 +1,163 @@
...
       }
     };
-    wsClient.addEventListener("message", handleMessage);
+    ws.addEventListener("message", handleMessage);
 
     function handleOpen() {
-      wsClient.send(JSON.stringify({ type: "init", projectId }));
-      wsClient.removeEventListener("open", handleOpen);
+      const message =
+        "sessionId" in options && options.sessionId
```
Comment (heikir) [inline chat-client.ts]: This is completely unreadable. Let's please stick to if's or extract it into a function.

File: types.ts [hunk 5]
```diff
@@ -10,6 +10,19 @@ export type ClientOptions = {
   defaultVariables?: Record<string, string>;
 };
 
+type IsNode = typeof globalThis extends { process: unknown } ? true : false;
```
Comment (heikir) [inline types.ts:13]: Is there a better way how to define that type?

**Review by heikir — APPROVED** [review]
LGTM!

**Review by christophwitzko — APPROVED** [review]
LGTM ✨ Great work!

File: chat-client.ts (hunk 2 above)
Comment (Gregoor) [reply chat-client.ts]: True, I made it a constant for now

File: chat-client.ts (hunk 3 above)
Comment (Gregoor) [reply chat-client.ts]: Unfortunately there is no error-event

File: chat-client.ts (hunk 4 above)
Comment (Gregoor) [reply chat-client.ts]: I think that's an opinion, but I feel less strongly about it. Replaced it with an if-else which I find a tad less readable

File: types.ts (hunk 5 above)
Comment (Gregoor) [reply types.ts:13]: I did not really find a better way!

File: chat-client.ts [hunk 6]
```diff
@@ -1,9 +1,11 @@
 import { ChatClientOptions } from "./types";
 
+const WEBSOCKET_URL = "wss://api.commonbase.com/chats";
```
Comment (christophwitzko) [inline chat-client.ts:3]: Should we put this in this file: https://github.com/commonbaseapp/sdk/blob/main/constants.ts

**Pull request merged**